### PR TITLE
[web] Include flags in custom events

### PR DESCRIPTION
**Pull request by AndyBitz**
Fixes FLA-1083

Follow up to and requires https://github.com/vercel/front/pull/31293
Ensures that the flags can be passed to the analytics script.

### CODE REVIEW

File: packages/web/src/server/index.ts [hunk 1]
```diff
@@ -98,12 +107,26 @@ export async function track(
...
+      flagsToReport = allFlags;
+    }
+
     const body = {
       o: origin,
       ts: new Date().getTime(),
       r: '',
       en: eventName,
       ed: props,
+      flags: {
```
Comment (dferber90) [inline packages/web/src/server/index.ts:127]: should this just be `f` to keep it shorter?

Comment (AndyBitz) [reply packages/web/src/server/index.ts:127]: Maybe. I think the reason the other keys are so short is a historical one, so I'd rather be more descriptive, but will do whatever the Analytics team sees fit here.

File: packages/web/src/server/index.ts [hunk 2]
```diff
@@ -98,12 +107,26 @@ export async function track(
 
     const url = new URL(origin);
 
+    let flagsToReport: Record<string, unknown> = {};
+    const allFlags = requestContext?.flags?.getValues();
+    if (options?.flags && allFlags) {
```
Comment (dferber90) [inline packages/web/src/server/index.ts:112]: Does this mean we made the decision that flags will only get tracked when specified explicitly? No auto-tracking of all known flags?

I don't have an opinion here yet, just curious if this was a conscious decision

**Review by dferber90 — APPROVED** [review]

File: packages/web/src/server/index.ts (hunk 2 above)
Comment (AndyBitz) [reply packages/web/src/server/index.ts:112]: No, check the `else if` closure here. We'll default to all flags and let the API decide by default. The `options.flags` only provide a way to be more specific when needed.

Comment (dferber90) [reply packages/web/src/server/index.ts:112]: Oh, totally missed that!

If a user wanted to opt out of tracking any flags by default, and only opt certain `track` calls into tracking some flags, would this work already? I guess we're adding the global setting later on?

File: packages/web/src/server/index.ts [hunk 3]
```diff
@@ -98,12 +107,26 @@ export async function track(
 
     const url = new URL(origin);
 
+    let flagsToReport: Record<string, unknown> = {};
+    const allFlags = requestContext?.flags?.getValues();
```
Comment (dferber90) [inline packages/web/src/server/index.ts:111]: nit for readability

I'd rename:
- `flagsToReport` -> `flagValuesToReport`
- `allFlags` -> `allFlagValues`

This avoids using the ambiguous term "flag" which can refer to the key of a flag, or the key/value of a flag.

File: packages/web/src/server/index.ts [hunk 4]
```diff
@@ -98,12 +107,26 @@ export async function track(
...
+    }
+
     const body = {
       o: origin,
       ts: new Date().getTime(),
       r: '',
       en: eventName,
       ed: props,
+      flags: {
+        plainValues: flagsToReport,
```
[the remark this answers is not in the document]
Comment (AndyBitz) [reply packages/web/src/server/index.ts:128]: `flags` here must support 3 properties: `keys`, `plainValues`, and `encryptedValues` for the client. I don't think it's a good idea to change it for the server and support 2 different schemas for validation on the ingest API.

File: packages/web/src/generic.ts [hunk 5]
```diff
@@ -106,7 +107,7 @@ function track(
   }
 
   if (!properties) {
-    window.va?.('event', { name });
+    window.va?.('event', { name, options });
```
Comment (feugy) [inline packages/web/src/generic.ts:110]: **suggestion(blocking):** While the user-facing property is `flags`, the analytics script expect [`flagsKeys`](https://github.com/vercel/front/pull/31293/files#diff-fe13bc40f2a9c7c1cbb5ae022240a9ae45689228be8f30f8eff3fd13bac828b6R19)

File: packages/web/src/generic.ts [hunk 6]
```diff
@@ -118,6 +119,7 @@ function track(
     window.va?.('event', {
       name,
       data: props,
+      options,
```
[the remark this answers is not in the document]
Comment (feugy) [reply packages/web/src/generic.ts:122]: same here, we'll need to map `flags` to `flagKeys` here as well

File: packages/web/src/server/index.ts [hunk 7]
```diff
@@ -98,12 +107,26 @@ export async function track(
...
 
+    let flagsToReport: Record<string, unknown> = {};
+    const allFlags = requestContext?.flags?.getValues();
+    if (options?.flags && allFlags) {
+      options.flags.forEach((key) => {
+        flagsToReport[key] = allFlags[key];
+      });
+    } else if (allFlags) {
+      // Default to all flags. The ingest endpoint will truncate them.
+      flagsToReport = allFlags;
```
Comment (feugy) [inline packages/web/src/server/index.ts:118]: **suggestion(blocking):** as discussed, we should remove this code. If users haven't specified a list of tracked keys, we should not include anything.

**Review by feugy — COMMENTED** [review]
In addition to the small changes here, you'll need to adjust the react component so it takes a list of tracked keys, and populates the script data attribute.
The script will then use them for page views.

To move further, you can bump the package version as alpha, bounded to your script on the preview deployment. This will allow testing it end to end.

File: packages/web/src/server/index.ts (hunk 7 above)
Comment (tobiaslins) [reply packages/web/src/server/index.ts:118]: I'm fine with sending all here (given its happening on the API and will see less flags here anyways?)

File: packages/web/src/server/index.ts [hunk 8]
```diff
@@ -33,7 +34,7 @@ const logPrefix = '[Vercel Web Analytics]';
 export async function track(
   eventName: string,
   properties?: Record<string, AllowedPropertyValues>,
-  context?: Context
+  options?: Options
```
Comment (feugy) [inline packages/web/src/server/index.ts:37]: **praise:** thanks for renaming this parameter in something more aligned with the other existing methods

**Review by feugy — COMMENTED** [review]
I'm good with it.

Let's not merge it yet, so we we could ship and release urgent fixes until we decide to release the feature in this PR.

File: packages/web/src/types.ts [hunk 9]
```diff
@@ -25,12 +40,14 @@ export interface AnalyticsProps {
   endpoint?: string;
 
   dsn?: string;
+
+  setDefaultProps?: () => DefaultProps;
```
[the remark this answers is not in the document]
Comment (AndyBitz) [reply packages/web/src/types.ts]: Renamed it and added a comment 👍

File: packages/web/src/generic.ts [hunk 10]
```diff
@@ -89,7 +89,11 @@ function inject(
  */
 function track(
   name: string,
-  properties?: Record<string, AllowedPropertyValues>
+  properties?: Record<string, AllowedPropertyValues>,
+  options?: {
+    flagKeys?: string[];
+    flags?: Record<string, boolean | number | string>;
```
Comment (AndyBitz) [inline packages/web/src/generic.ts:95]: @tobiaslins you've added this here, but the script does not seem to support it. Should we remove it again?

Comment (AndyBitz) [reply packages/web/src/generic.ts]: Yea, removed it

**Pull request merged**